S6 t2/create frontend displaying upcoming meetings

catai9 opened the pull request:
**Scrum Board Ticket:** S6 T2: Create the front end interface for displaying all upcoming appointments on the dashboard.

**Story:** Story 6 Meeting/Appointment Descriptions

**Description**: Added an AppointmentList component to display the upcoming appointments on the Home page and also installed dependencies to run the unit tests and sass. This PR also involves updating the navigation bar to display softer colours for worker and student users (as per the TA feedback).

**Testing Criteria:** Tested code by running locally and ensuring that the Home page displayed as expected for both the worker and student users. I also ran the automated tests codes and ensured that there was 100% file coverage (and passing tests) for all the server tests. I also ran the unit tests in the front end and ensured 100% file coverage (and passing tests) for the AppointmentList component. The coverage for the other tests are in scope for another ticket and not set up yet (hence why less than 100% file coverage for those files). 

**How Reviewers can Test Code:**
1. Run the code locally and ensure no complications when starting up (run the command `npm run-script start:dev` in the root folder OR run the command `npm start` if your node and npm versions are above 12.10.0 for node nad 6.10.3 for npm).
a. Use the credentials `username: johndoe@gmail.com and password: a1234` to view the upcoming appointments. The following window should appear: 
![image](https://user-images.githubusercontent.com/32720322/97810605-12479600-1c43-11eb-9eaa-ef4fcc12a3fe.png)
b. Use the credentials `username: joshuabrooks@gmail.com and password: j1234` to view the upcoming appointments. The following window should appear:  
![image](https://user-images.githubusercontent.com/32720322/97810616-3014fb00-1c43-11eb-99af-bbfb6e2693e3.png)

2. Run the automated server tests locally and ensure all tests pass and have 100% file coverage. 
a. Navigate inside the `server` folder and run `npm test`. 
b. Navigate inside the `react-ui` folder and run `npm test`. All tests should pass and the AppointmentList component should have 100% file coverage (the rest are not set up yet and are in the scope for another ticket).

catai9 commented on react-ui/src/App.js:
```diff
@@ -28,7 +28,7 @@ class App extends Component {
           {/* <LogIn /> */}
         </Route>
-        <Route exact path="/dashboard/:email/:type/:name" component={Dashboard}>
+        <Route exact path="/dashboard/:email/:userType/:firstName/:personId/:accessToken" component={Dashboard}>
```
Please note that additional props were needed from the login page. Although this method is not a secure way of passing sensitive parameters such as accessToken, adjustments such as implementing a redux store or leveraging another method is out of scope for this ticket and will be considered for a future sprint.

sehersarin replied:
This looks good, thank you for the reasoning.

catai9 commented on react-ui/src/__snapshots__/App.test.js.snap:
```diff
@@ -0,0 +1,119 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
snapshot code can be effectively ignored as they are created automatically and used only for testing comparisons.

catai9 commented on react-ui/src/App.test.js:
```diff
@@ -0,0 +1,11 @@
+import React from "react";
```
Please note that this PR only sets up the initial test for the App component. Additional tests are in scope for another ticket and out of scope for this ticket.

sehersarin replied:
Great, thanks for clarifying the scope of this ticket.

catai9 commented on react-ui/src/constants/userTypes.json:
```diff
@@ -0,0 +1,4 @@
+{
```
A file was created to standardize the userTypes across the front end components. Unfortunately, this constants file cannot be in the root folder (for sharing across the server and react-ui codebase) due to limitations with React.

sehersarin replied:
Understood - this does make sense.

catai9 commented on server/package.json:
```diff
@@ -11,8 +11,6 @@
     "dotenv": "^4.0.0",
     "express": "^4.16.2",
     "joi": "^17.2.1",
-    "lodash": "^4.17.20",
```
These dependencies were moved to the main folder to reduce unnecessary repetition as they are used in both the react-ui and server codebases.

catai9 commented on server/models/data/Student.js:
```diff
@@ -7,7 +7,7 @@ class Student {
         this.firstName = studentData.first_name;
         this.lastName = studentData.last_name;
         this.email = studentData.email;
-        this.type = UserTypes.student;
```
Please note that this change was due to the conflicting issue with the worker type attribute.

catai9 commented on react-ui/src/components/Layouts/Profile.js:
```diff
@@ -1,7 +1,7 @@
 import React, { Component, Fragment } from "react";
 import Title from "../Title"
 
-class Page2 extends Component {
```
This file was renamed to increase clarity and avoid user confusion (as mentioned in the TA demo) but no actual work was done for this page for this PR (out of scope).

sehersarin replied:
Yes, I agree - thank you for renaming this file, will definitely help avoid any confusion.

catai9 commented on react-ui/src/components/Layouts/Home.js:
```diff
@@ -1,13 +1,16 @@
+import AppointmentList from "../AppointmentList/AppointmentList";
 
-class IndexDashboard extends Component {
+class Home extends Component {
```
Renamed this page to Home to increase clarity as IndexDashboard may be confusing.

sehersarin replied:
Sounds good to me.

catai9 commented on react-ui/src/components/Dashboard/Dashboard.js:
```diff
@@ -0,0 +1,95 @@
+import React, { Component, Fragment, Router } from "react";
```
Please note that I moved Dashboard to its own folder as it is different from the typical Layouts and is used across all components (after login).

sehersarin replied:
That makes sense!

sehersarin commented:
Hi Amy, took a look through your PR and tried running the code however, I am getting this error - do you know why this could be? (I ran the npm install command and then npm start in the react-ui folder.)

<img width="1055" alt="Screen Shot 2020-11-01 at 1 57 33 PM" src="https://user-images.githubusercontent.com/43328744/97811727-66a24400-1c4a-11eb-9883-8a23079269d7.png">

sehersarin commented:
Additionally, I see that there are 3 failing checks for CI - would you please be able to take a look.

mtatran commented:
<img width="740" alt="Screen Shot 2020-11-01 at 2 46 06 PM" src="https://user-images.githubusercontent.com/30379133/97812692-0b278480-1c51-11eb-80e4-746540880729.png">
 Jest tests are failing

sehersarin commented:
> Hi Amy, took a look through your PR and tried running the code however, I am getting this error - do you know why this could be? (I ran the npm install command and then npm start in the react-ui folder.)
> 
> <img alt="Screen Shot 2020-11-01 at 1 57 33 PM" width="1055" src="https://user-images.githubusercontent.com/43328744/97811727-66a24400-1c4a-11eb-9883-8a23079269d7.png">

Update: Was able to fix this by doing an npm install in the root folder.

sehersarin commented:
Have verified the UI for both login credentials a) and b) and they look good to me! 

However, when I try to run test cases there are some errors or low code coverage. 
1) When running npm test in react-ui folder this is what I see (screenshot below)

<img width="738" alt="Screen Shot 2020-11-01 at 2 58 44 PM" src="https://user-images.githubusercontent.com/43328744/97812984-25626200-1c53-11eb-9f36-3bb924b98d1f.png">


2.) When running npm test in server folder it shows 3 tests and 3 test suites fail.

<img width="647" alt="Screen Shot 2020-11-01 at 2 59 54 PM" src="https://user-images.githubusercontent.com/43328744/97812986-298e7f80-1c53-11eb-8000-30e15f911683.png">

catai9 commented:
Hi Seher, 
Glad to hear that worked! 
1. I fixed the failing checks by deleting and reinstalling the dependencies again. 
2. I also updated the test files in the server folder to account for the change in `userType`. All server tests should be passing now. 
3. The red lines on the front end tests are not errors - they signify uncovered lines and low code coverage (which is in scope for another task). In your screenshot, you can see that the AppointmentList component contains 100% code coverage. As shown below, these lines (located at the bottom) indicate if tests pass:  
![image](https://user-images.githubusercontent.com/32720322/97813206-89395a80-1c54-11eb-9588-b4d5cf262396.png).

catai9 commented:
> 
> 
> <img alt="Screen Shot 2020-11-01 at 2 46 06 PM" width="740" src="https://user-images.githubusercontent.com/30379133/97812692-0b278480-1c51-11eb-80e4-746540880729.png">
> Jest tests are failing

Hi Melissa, you might need to do `npm install` again to make sure that all of your dependencies are up-to-date.

sehersarin commented:
Hi Amy, 

Thank you for making all of the requested changes and fixing the errors for the server tests. Have verified the UI and code coverage and everything looks good to me. I'm ready to approve and merge the PR to the master branch.
